### PR TITLE
Locally-working Zupoll for Edge City

### PR DESCRIPTION
**Pull request by robknight**

### CODE REVIEW

File: apps/zupoll-server/src/util/verify.ts [hunk 1]
```diff
@@ -196,10 +208,20 @@ export async function verifySignatureProof(
...
+
+async function verifyGenericIssuanceRootValidity(
+  baseUrl: string,
+  root: string,
+): Promise<boolean> {
+  const url = baseUrl + "/valid/" + root;
+  const response = await fetch(url);
+  const result = await response.json();
+  return result.valid;
+}
```
Comment (ichub) [inline apps/zupoll-server/src/util/verify.ts:227]: I am not sure that this doesn't create security problems. what stops someone from passing in some arbitrary semaphore group? I suppose that whether this url is actually the edge cities pipeline semaphore group url must be checked on the way to this function. Nevertheless, may be nice to check that here too.

**Review by ichub — APPROVED** [review]
lfg

File: apps/zupoll-client/pages/index.tsx [hunk 2]
```diff
@@ -38,6 +38,8 @@ export default function Index() {
   if (!isLoading && !loginState) {
     content = (
       <LoginScreen
+        title="This app lets Zupass users vote anonymously."
+        visibleLoginOptions={[]}
```
[the remark this answers is not in the document]
Comment (robknight) [reply apps/zupoll-client/pages/index.tsx:42]: Ugh, I renamed this variable and that changed the semantics. There is no definitive list of all of the login options anywhere, so I was using the empty array to indicate "don't filter the login options". I'll come up with a better way to do this.

File: apps/zupoll-client/pages/denver.tsx [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+import { useSavedLoginState } from "../src/useLoginState";
+import { removeQueryParameters } from "../src/util";
+
+export default function Denver() {
+  const router = useRouter();
+
+  useEffect(() => {
+    // Check if the query parameter exists
+    if (router.query.tgWebAppStartParam) {
+      console.log(`got start param`, router.query.tgWebAppStartParam);
```
[the remark this answers is not in the document]
Comment (robknight) [reply apps/zupoll-client/pages/denver.tsx:17]: fwiw this was directly copied from index.tsx

File: apps/zupoll-client/components/main/MainScreen.tsx [hunk 4]
```diff
@@ -96,36 +96,36 @@ export function MainScreen({
 
       {(loginState.config.name === LoginConfigurationName.ZUZALU_ORGANIZER ||
         loginState.config.name ===
-          LoginConfigurationName.ZUZALU_PARTICIPANT) && (
-        <BallotListContainer>
-          <TitleContainer>
-            <H1>Organizer Polls</H1>
-            <p>Official ballots from Zuconnect organizers</p>
-          </TitleContainer>
+        LoginConfigurationName.ZUZALU_PARTICIPANT) && (
```
[the remark this answers is not in the document]
Comment (robknight) [reply apps/zupoll-client/components/main/MainScreen.tsx:99]: I don't think I know how to do that. I can see that it's changing things as I go, but I haven't changed either the prettier config or the vscode config.

File: apps/zupoll-server/src/bot.ts [hunk 5]
```diff
@@ -112,7 +112,7 @@ export async function startBot(context: ApplicationContext): Promise<void> {
       ctx.reply(`Zupoll`, {
         reply_markup: new InlineKeyboard().url(
           `Zupoll`,
-          `https://t.me/zupoll_prod_bot/poll`
+          process.env.BOT_ZUPOLL_LINK ?? `https://t.me/zupoll_prod_bot/poll`,
```
[the remark this answers is not in the document]
Comment (robknight) [reply apps/zupoll-server/src/bot.ts:115]: How do we make it required?

File: apps/zupoll-server/src/util/verify.ts (hunk 1 above)
Comment (robknight) [reply apps/zupoll-server/src/util/verify.ts]: I don't think it ought to work like this, but this is a direct copy of the existing function for accessing Zupass Semaphore groups, with a slight change to the URL structure. It's no more or less secure than that.

In practice this function is only called with inputs that are among the hard-coded Semaphore group URLs, so it's not practically possible to insert arbitrary URLs.

File: apps/zupoll-client/components/main/MainScreen.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (robknight) [reply apps/zupoll-client/components/main/MainScreen.tsx]: I've copied the prettier and vscode settings from the Zupass repo, and have reformatted everything. This has eliminated some changes and added some new ones, but at least the formatting is nicer now.

**Pull request merged**